compiler-plugin/src/main/kotlin/arrow/meta/quotes/ClassOrObject.kt

```diff
@@ -13,12 +13,150 @@ import org.jetbrains.kotlin.psi.psiUtil.getValueParameters
+ * compiler by taking advantage of typeclass checks.
+ *
+ * In the [LensPlugin], we check to see if a type is a product type.  To add context, the cardinality of a product
+ * type is the product of their cardinalities:
+ *
+ * ```
+ *      |(a, b)| = |a| x |b|
+ * ```
+ *
+ * In particular, the [LensPlugin] is passing a function to whatever class or object type is a product type:
+ *
+ * ```
+ *  val Meta.lenses: Plugin
+ *    get() =
+ *      "lenses" {
+ *        meta {
+ *          classOrObject(::isProductOrType) { c -> ... }
+ *        }
+ *      }
+ * ```
+ *
+ * Where the function type for [::isProductType] is written as:
+ *
+ * ```
+ * fun isProductType(ktClass: KtClass): Boolean =
+ *      ktClass.isData() &&
+ *        ktClass.primaryConstructorParameters.isNotEmpty() &&
+ *        ktClass.primaryConstructorParameters.all { !it.isMutable } &&
+ *        ktClass.typeParameters.isEmpty()
+ * ```
+ *
+ * For a [KtClass] to be a product type, we check to see if the [KtClass] is an ADT, and that the cardinality of
+ * the constructor parameters is greater than one and non-mutable. For the [LensPlugin], we've just demonstrated how we
+ * can check for the product type of a [KtClass] by access the properties made available by the Kotlin PSI change.
+ *
+ * While it is not necessary to build [KtClass] predicates for [Meta.classOrObject] does not necessarily have to serve
+ * as a check for functional proofs in your compiler plugin.  It is relevant to highlight that the `match` predicate
+ * written gives access to the Kotlin PSI when analyzing written code via the written compiler plugin.
+ *
+ * The second parameter `map` is a function that allows the resulting action from matching on the transformation at
+ * the PSI level. Let us example another example plugin:
+ *
+ * ```kotlin:ank:silent
+ * import arrow.meta.Meta
+ * import arrow.meta.Plugin
+ * import arrow.meta.invoke
+ * import arrow.meta.quotes.Transform
+ * import arrow.meta.quotes.classOrObject
+ * import org.jetbrains.kotlin.psi.KtClass
+ * import com.intellij.psi.PsiElement
+ *
+ * val Meta.example: Plugin
+ *   get() =
+ *     "Example" {
+ *       meta(
+ *         /** Intercepts all classes named 'Test' **/
+ *         classOrObject({ name == "Test" }) { classOrObject ->
+ *           Transform.replace<KtClass>(
+ *             replacing = classOrObject,
+ *             newDeclaration =
+ *               """|$`@annotationEntries` $kind $name $`(typeParameters)` $`(valueParameters)` : $supertypes"} {
+ *                  |  $body
+ *                  |  fun void test(): Unit =
+ *                  |    println("Implemented by ΛRROW Meta!")
+ *                  |}
+ *                  |""".`class`.synthetic
+ *           )
+ *         }
+ *       )
+ *     }
+ * ```
+ *
+ * After analyzing the PSI elements available, we pass a resulting [KtClass] matching the predicate (in our case,
```

Comment:
I guess what I wrote above in https://github.com/arrow-kt/arrow-meta/pull/87/files#r344923142 is similar to this so feel free to discard or incorporate whatever you think is worth it

Reply:
I had a rather messy rebase on this particular one since it's been sitting for a couple days - I definitely have to go back to double check stuff like this!